cmd/caa-log-checker/main.go

```diff
@@ -79,19 +82,39 @@ func checkIssuances(scanner *bufio.Scanner, checkedMap map[string][]time.Time, s
 		}
 
 		var badNames []string
+		var minTimeError time.Duration = math.MaxInt64
+
 		for _, name := range ie.Names {
 			nameOk := false
+
 			for _, t := range checkedMap[name] {
-				if t.Before(ie.issuanceTime) && t.After(ie.issuanceTime.Add(-8*time.Hour)) {
+				validStart := ie.issuanceTime.Add(-8 * time.Hour)
+				validEnd := ie.issuanceTime
+				if t.After(validStart) && t.Before(validEnd.Add(timeTolerance)) {
 					nameOk = true
+				} else {
+					// If the check didn't pass, calculate how much tolerance we'd need for it to pass, to make it easier to diagnose log timestamp desync.
+					if !t.Before(validEnd) {
+						timeError := t.Sub(validEnd)
+						// ...however only if its <1h, otherwise it's probably not a match
+						if timeError < minTimeError && timeError < maxTimeTolerance {
```

Comment:
It's not clear to me why `maxTimeTolerance` needs to be checked here. If the goal is that at the end of the loop, `minTimeError` contains the smallest delta such that there was a CAA check for `name` within delta seconds after the issuance log timestamp, it seems unnecessary to check the `maxTolerance` to achieve that. Also, it's a bit confusing that the `maxTimeTolerance` const seems to have two roles: One, as a limit on the value of the `-time-tolerance` flag, and two, as a clamp here.

Also, it seems like the `minTimeError` is a property of each name, and you want to calculate the maximum of it across all names. For instance, if you had a cert for `[a, b]`, and a's CAA check had a log timestamp 120 ms after the issuance log timestamp, and b's CAA check had a log timestamp 200 ms after the issuance log timestamp, and `-time-tolerance` was 90ms, you would want to report a `timeError` for that issuance of 200 ms.

Reply:
The purpose of `maxTimeTolerance` is to avoid spurious potential matches that are very far away from the issuance being checked. For example, consider someone who renews their certs once a day (and we somehow have CAA logs into the future, though the same thing can happen within a single day) and we fail to do a CAA check in only one of those instances. The tool will continue scanning forward into the logs and find the CAA check that was done the next day, and then report `timeError=86400.000` in the log, which is just a false positive.

The flag is checked against `maxTimeTolerance` because if `timeTolerance > maxTimeTolerance` then this check has no effect and it'll always report `timeError=NaN`. Thinking again, I'm not sure why I did this instead of just comparing against `timeTolerance + maxTimeTolerance` or something similar. I'll change this unless you can think of any reason why that would be bad.

You're correct about `minTimeError`, you'd want the maximum of the minimum times. I think I meant to handle multiple names but probably got confused about which direction you should min/max towards. I think what I'll do instead is make `timeError` a list parallel to badName so you get the value for each individual name.